lms/static/scripts/frontend_apps/components/BookPicker.js

```diff
@@ -0,0 +1,123 @@
+import { LabeledButton } from '@hypothesis/frontend-shared';
+import { createElement } from 'preact';
+import { useCallback, useEffect, useState } from 'preact/hooks';
+
+import { fetchBooks, fetchChapters } from '../utils/api';
+
+import BookList from './BookList';
+import ChapterList from './ChapterList';
+import Dialog from './Dialog';
+import ErrorDisplay from './ErrorDisplay';
+
+/**
+ * @typedef {import('../api-types').Chapter} Chapter
+ * @typedef {import('../api-types').Book} Book
+ *
+ * @typedef BookPickerProps
+ * @prop {string} authToken
+ * @prop {() => void} onCancel
+ * @prop {(b: Book, c: Chapter) => void} onSelectBook
+ */
+
+/**
+ * A dialog that allows the user to select a book and chapter to use in an assignment.
+ *
+ * The list of available books is fetched from VitalSource, but this component
+ * could be adapted to work with other book sources in future.
+ *
+ * @param {BookPickerProps} props
+ */
+export default function BookPicker({ authToken, onCancel, onSelectBook }) {
+  const [bookList, setBookList] = useState(/** @type {Book[]|null} */ (null));
+  const [chapterList, setChapterList] = useState(
+    /** @type {Chapter[]|null} */ (null)
+  );
+  const [book, setBook] = useState(/** @type {Book|null} */ (null));
+  const [chapter, setChapter] = useState(/** @type {Chapter|null} */ (null));
+  const [step, setStep] = useState(
+    /** @type {'select-book'|'select-chapter'} */ ('select-book')
+  );
+  const [error, setError] = useState(/** @type {Error|null} */ (null));
```

Comment:
I don't think its a problem at present. The thing I would pay more attention to rather than the number of state values is whether there is redundancy (eg. one state value always implying another) or whether the state structure makes it possible to represent states that should be impossible. In that case some refactoring is called for.

You could potentially push some of this state down into the `BookList` and `ChapterList` components and let them "own" the book / chapter list. If so however you then have an issue that you'd want to avoid re-fetching the list if we added a "Back" button to the dialog in future (that is, go back from the chapter selection screen to the book selection screen).